README.md
# Backbone webStorage Adapter v0.2.0

This is simply a fork from the excellent [Backbone.localStorage](https://github.com/jeromegn/Backbone.localStorage) by Jerome Gravel-Niquet and made compatible with both window.localStorage and window.sessionStorage as a drop-in replacement for Backbone.Sync() to handle saving to a localStorage or sessionStorage database.

## Usage

Include Backbone.webStorage after having included Backbone.js:

```html
<script type="text/javascript" src="backbone.js"></script>
<script type="text/javascript" src="backbone.webStorage.js"></script>
```

Create your collections like so:

```javascript
window.SomeCollection = Backbone.Collection.extend({
  
  // For localStorage use:
  localStorage: new Backbone.WebStorage("SomeCollection", "localStorage"), // Unique name within your app.
  // For sessionStorage use:
  sessionStorage: new Backbone.WebStorage("SomeCollection", "sessionStorage"), // Unique name within your app.
  
  // ... everything else is normal.
  
});
```
For furher usage instructions, see [Backbone.localStorage](https://github.com/jeromegn/Backbone.localStorage).

## Acknowledgments

- [Jerome Gravel-Niquet](https://github.com/jeromegn): Full credits to Jerome for the original Backbone.localStorage solution.

## License

Licensed under MIT license

Copyright (c) 2015 Glen Somerville
Copyright (c) 2010 Jerome Gravel-Niquet

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Changelog

## 0.2.0 - February 24, 2015

- Renamed to webStorage and made both sessionStorage and localStorage possible

## 0.1.1 - February 17, 2015

- Tweaked package description

## 0.1 - February 17, 2015

- Forked from the excellent [Backbone.localStorage](https://github.com/jeromegn/Backbone.localStorage) by Jerome Gravel-Niquet and made compatible with window.sessionStorage

spec/webStorage_spec.js


describe("Backbone.webStorage", function(){

  var attributes = {
    string: "String",
    string2: "String 2",
    number: 1337
  };

  var onError = function (model, resp, options) {
    throw new Error(resp);
  };

  describe("localStorage on a Collection", function(){

    var Model = Backbone.Model.extend({
      defaults: attributes
    });

    var Collection = Backbone.Collection.extend({
      model: Model,
      localStorage: new Backbone.WebStorage("collectionStore", "localStorage")
    });

    var collection = new Collection();

    // bind error handler
    before(function(){
      collection.on('error', onError);
    })

    // Clean up before starting
    before(function(){
      collection.localStorage._clear();
    });

    before(function(){
      collection.fetch();
    });

    it("should default to use `localSync`", function(){
      assert.equal(Backbone.getSyncMethod(collection), Backbone.localSync);
    });

    it("should use `ajaxSync` when option is passed", function(){
      assert.equal(Backbone.getSyncMethod(collection, { ajaxSync: true }), Backbone.ajaxSync);
    });

    it("should use `localSync` when other options are passed", function(){
      assert.equal(Backbone.getSyncMethod(collection, { ajaxSync: false, data: {}, silent: true }), Backbone.localSync);
    });

    it("should initially be empty", function(){
      assert.equal(collection.length, 0);
    });


    describe("create", function(){

      var model;

      before(function(){
        model = collection.create({});
      });

      it("should have 1 model", function(){
        assert.equal(collection.length, 1);
      });

      it("should have a populated model", function(){
        var withId = _.clone(attributes);
        withId.id = model.id;
        assert.deepEqual(model.toJSON(), withId);
      });

      it("should have assigned an `id` to the model", function(){
        assert.isDefined(model.id);
      });

      it("should be saved to the localStorage", function(){
        assert.isNotNull(window.localStorage.getItem('collectionStore'+'-'+model.id));
      });

    });

    describe("get (by `id`)", function(){

      var model;

      before(function(){
        model = collection.create({});
      });

      it("should find the model with its `id`", function(){
        assert.equal(collection.get(model.id), model);
      });

    });

    describe("instances", function(){

      describe("save", function(){

        var model, model2;

        before(function(){
          model = collection.create({});
          model.save({string: "String 0"});
          collection.fetch()
        });

        it("should persist the changes", function(){
          assert.equal(model.get("string"), "String 0");
        });


        describe("with a new `id`", function(){

          before(function(){
            model2 = collection.create({});
            model2.save({id: 1});
            collection.fetch();
          });

          it("should have a new `id`", function(){
            assert.equal(model2.id, 1);
          });

          it("should have kept its old properties", function(){
            var withId = _.clone(attributes);
            withId.id = 1;
            assert.deepEqual(model2.toJSON(), withId);
          });

          it("should be saved in localStorage by new id", function() {
            assert.isNotNull(window.localStorage.getItem('collectionStore-1'));
          });

        });


      });

      describe("destroy", function(){

        var beforeFetchLength, afterFetchLength;

        before(function(){
          // Make sure there's at least items in there
          // ... can't rely on previous tests
          _(5).times(function(){
            collection.create()
          });
        });

        before(function(){
          _.each(collection.toArray(), function(model){
            model.destroy();
          });
          beforeFetchLength = collection.length;
        });

        before(function(){
          collection.fetch();
          afterFetchLength = collection.length;
        });

        it("should have removed all items from the collection", function(){
          assert.equal(beforeFetchLength, 0);
        });

        it("should have removed all items from the store", function(){
          assert.equal(afterFetchLength, 0);
        });

      });

      describe("with a different `idAttribute`", function(){

        var Model2 = Backbone.Model.extend({
          defaults: attributes,
          idAttribute: "_id"
        });

        var Collection2 = Backbone.Collection.extend({
          model: Model2,
          localStorage: new Backbone.WebStorage("collection2Store", "localStorage")
        });

        var collection2 = new Collection2();

        before(function(){
          collection2.create();
        });

        it("should have used the custom `idAttribute`", function(){
          assert.equal(collection2.first().id, collection2.first().get("_id"));
        });

      });

    });

  });

  describe("sessionStorage on a Collection", function(){

    var Model = Backbone.Model.extend({
      defaults: attributes
    });

    var Collection = Backbone.Collection.extend({
      model: Model,
      sessionStorage: new Backbone.WebStorage("collectionStore", "sessionStorage")
    });

    var collection = new Collection();

    // bind error handler
    before(function(){
      collection.on('error', onError);
    })

    // Clean up before starting
    before(function(){
      collection.sessionStorage._clear();
    });

    before(function(){
      collection.fetch();
    });

    it("should default to use `localSync`", function(){
      assert.equal(Backbone.getSyncMethod(collection), Backbone.localSync);
    });

    it("should use `ajaxSync` when option is passed", function(){
      assert.equal(Backbone.getSyncMethod(collection, { ajaxSync: true }), Backbone.ajaxSync);
    });

    it("should use `localSync` when other options are passed", function(){
      assert.equal(Backbone.getSyncMethod(collection, { ajaxSync: false, data: {}, silent: true }), Backbone.localSync);
    });

    it("should initially be empty", function(){
      assert.equal(collection.length, 0);
    });


    describe("create", function(){

      var model;

      before(function(){
        model = collection.create({});
      });

      it("should have 1 model", function(){
        assert.equal(collection.length, 1);
      });

      it("should have a populated model", function(){
        var withId = _.clone(attributes);
        withId.id = model.id;
        assert.deepEqual(model.toJSON(), withId);
      });

      it("should have assigned an `id` to the model", function(){
        assert.isDefined(model.id);
      });

      it("should be saved to the sessionStorage", function(){
        assert.isNotNull(window.sessionStorage.getItem('collectionStore'+'-'+model.id));
      });

    });

    describe("get (by `id`)", function(){

      var model;

      before(function(){
        model = collection.create({});
      });

      it("should find the model with its `id`", function(){
        assert.equal(collection.get(model.id), model);
      });

    });

    describe("instances", function(){

      describe("save", function(){

        var model, model2;

        before(function(){
          model = collection.create({});
          model.save({string: "String 0"});
          collection.fetch()
        });

        it("should persist the changes", function(){
          assert.equal(model.get("string"), "String 0");
        });


        describe("with a new `id`", function(){

          before(function(){
            model2 = collection.create({});
            model2.save({id: 1});
            collection.fetch();
          });

          it("should have a new `id`", function(){
            assert.equal(model2.id, 1);
          });

          it("should have kept its old properties", function(){
            var withId = _.clone(attributes);
            withId.id = 1;
            assert.deepEqual(model2.toJSON(), withId);
          });

          it("should be saved in sessionStorage by new id", function() {
            assert.isNotNull(window.sessionStorage.getItem('collectionStore-1'));
          });

        });


      });

      describe("destroy", function(){

        var beforeFetchLength, afterFetchLength;

        before(function(){
          // Make sure there's at least items in there
          // ... can't rely on previous tests
          _(5).times(function(){
            collection.create()
          });
        });

        before(function(){
          _.each(collection.toArray(), function(model){
            model.destroy();
          });
          beforeFetchLength = collection.length;
        });

        before(function(){
          collection.fetch();
          afterFetchLength = collection.length;
        });

        it("should have removed all items from the collection", function(){
          assert.equal(beforeFetchLength, 0);
        });

        it("should have removed all items from the store", function(){
          assert.equal(afterFetchLength, 0);
        });

      });

      describe("with a different `idAttribute`", function(){

        var Model2 = Backbone.Model.extend({
          defaults: attributes,
          idAttribute: "_id"
        });

        var Collection2 = Backbone.Collection.extend({
          model: Model2,
          sessionStorage: new Backbone.WebStorage("collection2Store", "sessionStorage")
        });

        var collection2 = new Collection2();

        before(function(){
          collection2.create();
        });

        it("should have used the custom `idAttribute`", function(){
          assert.equal(collection2.first().id, collection2.first().get("_id"));
        });

      });

    });

  });

  describe("localStorage on a Model", function(){

    var Model = Backbone.Model.extend({
      defaults: attributes,
      localStorage: new Backbone.WebStorage("modelStore", "localStorage")
    });

    var model = new Model();

    before(function(){
      model.localStorage._clear();
    });

    it("should use `localSync`", function(){
      assert.equal(Backbone.getSyncMethod(model), Backbone.localSync);
    });

    it("should use `ajaxSync` when option is passed", function(){
      assert.equal(Backbone.getSyncMethod(model, { ajaxSync: true }), Backbone.ajaxSync);
    });

    it("should use `localSync` when other options are passed", function(){
      assert.equal(Backbone.getSyncMethod(model, { ajaxSync: false, data: {}, silent: true }), Backbone.localSync);
    });

    describe("fetch", function(){
      it('should fire sync event on fetch', function(done) {
        var model = new Model(attributes);
        model.on('sync', function(){
          done();
        });
        model.fetch();
      });
    });

    describe("save", function(){

      before(function(){
        model.save();
        model.fetch();
      });

      it("should have assigned an `id` to the model", function(){
        assert.isDefined(model.id);
      });

      it("should be saved to the localStorage", function(){
        assert.isNotNull(window.localStorage.getItem('modelStore'+'-'+model.id));
      });

      describe("with new attributes", function(){

        before(function(){
          model.save({number: 42});
          model.fetch();
        });

        it("should persist the changes", function(){
          assert.deepEqual(model.toJSON(), _.extend(_.clone(attributes), {id: model.id, number: 42}));
        });

      });

      describe('fires events', function(){
        before(function(){
          this.model = new Model();
        });
        after(function(){
          this.model.destroy();
        });

        it('should fire sync event on save', function(done) {
          this.model.on('sync', function(){
            this.model.off('sync');
            done();
          }, this);
          this.model.save({foo: 'baz'});
        });
      });

    });

    describe("destroy", function(){

      before(function(){
        model.destroy();
      });

      it("should have removed the instance from the store", function(){
        assert.lengthOf(Model.prototype.localStorage.findAll(), 0);
      });

    });

  });

  describe("sessionStorage on a Model", function(){

    var Model = Backbone.Model.extend({
      defaults: attributes,
      sessionStorage: new Backbone.WebStorage("modelStore", "sessionStorage")
    });

    var model = new Model();

    before(function(){
      model.sessionStorage._clear();
    });

    it("should use `localSync`", function(){
      assert.equal(Backbone.getSyncMethod(model), Backbone.localSync);
    });

    it("should use `ajaxSync` when option is passed", function(){
      assert.equal(Backbone.getSyncMethod(model, { ajaxSync: true }), Backbone.ajaxSync);
    });

    it("should use `localSync` when other options are passed", function(){
      assert.equal(Backbone.getSyncMethod(model, { ajaxSync: false, data: {}, silent: true }), Backbone.localSync);
    });

    describe("fetch", function(){
      it('should fire sync event on fetch', function(done) {
        var model = new Model(attributes);
        model.on('sync', function(){
          done();
        });
        model.fetch();
      });
    });

    describe("save", function(){

      before(function(){
        model.save();
        model.fetch();
      });

      it("should have assigned an `id` to the model", function(){
        assert.isDefined(model.id);
      });

      it("should be saved to the sessionStorage", function(){
        assert.isNotNull(window.sessionStorage.getItem('modelStore'+'-'+model.id));
      });

      describe("with new attributes", function(){

        before(function(){
          model.save({number: 42});
          model.fetch();
        });

        it("should persist the changes", function(){
          assert.deepEqual(model.toJSON(), _.extend(_.clone(attributes), {id: model.id, number: 42}));
        });

      });

      describe('fires events', function(){
        before(function(){
          this.model = new Model();
        });
        after(function(){
          this.model.destroy();
        });

        it('should fire sync event on save', function(done) {
          this.model.on('sync', function(){
            this.model.off('sync');
            done();
          }, this);
          this.model.save({foo: 'baz'});
        });
      });

    });

    describe("destroy", function(){

      before(function(){
        model.destroy();
      });

      it("should have removed the instance from the store", function(){
        assert.lengthOf(Model.prototype.sessionStorage.findAll(), 0);
      });

    });

  });

  describe("localStorage Error handling", function(){

    var Model = Backbone.Model.extend({
      defaults: attributes,
      localStorage: new Backbone.WebStorage("modelStore", "localStorage")
    });

    describe("private browsing", function(){

      var model = new Model()
        , oldSetItem = window.localStorage.setItem
        , oldStorageSize = model.localStorage._storageSize
        , error;

      before(function(done){
        model.localStorage._clear();

        // Patch browser conditions for private error.
        model.localStorage._storageSize = function(){ return 0; };
        window.localStorage.setItem = function(){
          var error = new Error();
          error.code = DOMException.QUOTA_EXCEEDED_ERR;
          throw error;
        };

        model.save(attributes, {
          error: function(model, err){
            error = err;
            done();
          }
        })
      });

      it("should return the error in the error callback", function(){
        assert.equal(error, "Private browsing is unsupported");
      });

      it('should throw an error event', function(done){
        model.on('error', function() {
          done();
        });
        model.save();
      });

      after(function(){
        // Unwrap patches.
        model.localStorage._storageSize = oldStorageSize;
        window.localStorage.setItem = oldSetItem;
      })

    });

  });

  describe("sessionStorage Error handling", function(){

    var Model = Backbone.Model.extend({
      defaults: attributes,
      sessionStorage: new Backbone.WebStorage("modelStore", "sessionStorage")
    });

    describe("private browsing", function(){

      var model = new Model()
        , oldSetItem = window.sessionStorage.setItem
        , oldStorageSize = model.sessionStorage._storageSize
        , error;

      before(function(done){
        model.sessionStorage._clear();

        // Patch browser conditions for private error.
        model.sessionStorage._storageSize = function(){ return 0; };
        window.sessionStorage.setItem = function(){
          var error = new Error();
          error.code = DOMException.QUOTA_EXCEEDED_ERR;
          throw error;
        };

        model.save(attributes, {
          error: function(model, err){
            error = err;
            done();
          }
        })
      });

      it("should return the error in the error callback", function(){
        assert.equal(error, "Private browsing is unsupported");
      });

      it('should throw an error event', function(done){
        model.on('error', function() {
          done();
        });
        model.save();
      });

      after(function(){
        // Unwrap patches.
        model.sessionStorage._storageSize = oldStorageSize;
        window.sessionStorage.setItem = oldSetItem;
      })

    });

  });

});

describe("Without Backbone.webStorage", function(){

  describe("on a Collection", function(){
    var Collection = Backbone.Collection.extend()
      , collection = new Collection();

    it("should use `ajaxSync`", function(){
      assert.equal(Backbone.getSyncMethod(collection), Backbone.ajaxSync);
      assert.equal(Backbone.getSyncMethod(collection, { ajaxSync: false }), Backbone.ajaxSync);
    });
  });

  describe("on a Model", function(){
    var Model = Backbone.Model.extend()
      , model = new Model();

    it("should use `ajaxSync`", function(){
      assert.equal(Backbone.getSyncMethod(model), Backbone.ajaxSync);
      assert.equal(Backbone.getSyncMethod(model, { ajaxSync: false }), Backbone.ajaxSync);
    });
  });

});


// For some reason this is not ran when viewed in a browser
// but it is ran when using `mocha-phantomjs`.
describe("AMD", function(){

  require.config({
    paths: {
      jquery: "support/jquery",
      underscore: "support/underscore",
      backbone: "support/backbone",
      webStorage: "../backbone.webStorage"
    }
  });

  var WebStorage;

  before(function(done){
    require(["webStorage"], function(WS){
      WebStorage = WS;
      done()
    });
  });

  it("should be the same as the non-amd usage", function(){
    assert.equal(Backbone.WebStorage, WebStorage)
  });

});


spec/webStorage_commonjs_spec.js
describe("Backbone.webStorage in CommonJS environment", function() {

  var WebStorage = require("../backbone.webStorage");
  var Backbone = require("backbone");

  it("should be the same as the non-CommonJS usage", function(){
    assert.equal(Backbone.WebStorage, WebStorage);
  });
});


backbone.webStorage-min.js
/**
 * Backbone webStorage Adapter
 * Version 0.2.0
 *
 * https://github.com/glensomerville/Backbone.webStorage
 */(function(a,b){typeof exports=="object"&&typeof require=="function"?module.exports=b(require("backbone")):typeof define=="function"&&define.amd?define(["backbone"],function(c){return b(c||a.Backbone)}):b(Backbone)})(this,function(a){function b(){return((1+Math.random())*65536|0).toString(16).substring(1)}function c(){return b()+b()+"-"+b()+"-"+b()+"-"+b()+"-"+b()+b()+b()}function d(a){return a===Object(a)}function e(a,b){var c=a.length;while(c--)if(a[c]===b)return!0;return!1}function f(a,b){for(var c in b)a[c]=b[c];return a}function g(a,b){if(a==null)return void 0;var c=a[b];return typeof c=="function"?a[b]():c}return a.WebStorage=window.Store=function(a,b){if(!b)throw"Backbone.webStorage: Storage type not defined.";if(!this[b])throw"Backbone."+b+": Environment does not support "+b+".";this.name=a,this.storage=b,this.serializer={serialize:function(a){return d(a)?JSON.stringify(a):a},deserialize:function(a){return a&&JSON.parse(a)}};var c=this[this.storage]().getItem(this.name);this.records=c&&c.split(",")||[]},f(a.WebStorage.prototype,{save:function(){this[this.storage]().setItem(this.name,this.records.join(","))},create:function(a){return!a.id&&a.id!==0&&(a.id=c(),a.set(a.idAttribute,a.id)),this[this.storage]().setItem(this._itemName(a.id),this.serializer.serialize(a)),this.records.push(a.id.toString()),this.save(),this.find(a)},update:function(a){this[this.storage]().setItem(this._itemName(a.id),this.serializer.serialize(a));var b=a.id.toString();return e(this.records,b)||(this.records.push(b),this.save()),this.find(a)},find:function(a){return this.serializer.deserialize(this[this.storage]().getItem(this._itemName(a.id)))},findAll:function(){var a=[];for(var b=0,c,d;b<this.records.length;b++)c=this.records[b],d=this.serializer.deserialize(this[this.storage]().getItem(this._itemName(c))),d!=null&&a.push(d);return a},destroy:function(a){this[this.storage]().removeItem(this._itemName(a.id));var b=a.id.toString();for(var c=0,d;c<this.records.length;c++)this.records[c]===b&&this.records.splice(c,1);return this.save(),a},localStorage:function(){return localStorage},sessionStorage:function(){return sessionStorage},_clear:function(){var a=this[this.storage](),b=new RegExp("^"+this.name+"-");a.removeItem(this.name);for(var c in a)b.test(c)&&a.removeItem(c);this.records.length=0},_storageSize:function(){return this[this.storage]().length},_itemName:function(a){return this.name+"-"+a}}),a.WebStorage.sync=window.Store.sync=a.localSync=function(b,c,d){var e=g(c,"localStorage")||g(c.collection,"localStorage")||g(c,"sessionStorage")||g(c.collection,"sessionStorage"),f,h,i=a.$?a.$.Deferred&&a.$.Deferred():a.Deferred&&a.Deferred();try{switch(b){case"read":f=c.id!=undefined?e.find(c):e.findAll();break;case"create":f=e.create(c);break;case"update":f=e.update(c);break;case"delete":f=e.destroy(c)}}catch(j){j.code===22&&e._storageSize()===0?h="Private browsing is unsupported":h=j.message}return f?(d&&d.success&&(a.VERSION==="0.9.10"?d.success(c,f,d):d.success(f)),i&&i.resolve(f)):(h=h?h:"Record Not Found",d&&d.error&&(a.VERSION==="0.9.10"?d.error(c,h,d):d.error(h)),i&&i.reject(h)),d&&d.complete&&d.complete(f),i&&i.promise()},a.ajaxSync=a.sync,a.getSyncMethod=function(b,c){var d=c&&c.ajaxSync;return!d&&(g(b,"localStorage")||g(b.collection,"localStorage")||g(b,"sessionStorage")||g(b.collection,"sessionStorage"))?a.localSync:a.ajaxSync},a.sync=function(b,c,d){return a.getSyncMethod(c,d).apply(this,[b,c,d])},a.WebStorage});

backbone.webStorage.js
/**
 * Backbone webStorage Adapter
 * Version 0.2.0
 *
 * https://github.com/glensomerville/Backbone.webStorage
 */
(function (root, factory) {
  if (typeof exports === 'object' && typeof require === 'function') {
    module.exports = factory(require("backbone"));
  } else if (typeof define === "function" && define.amd) {
    // AMD. Register as an anonymous module.
    define(["backbone"], function(Backbone) {
      // Use global variables if the locals are undefined.
      return factory(Backbone || root.Backbone);
    });
  } else {
    factory(Backbone);
  }
}(this, function(Backbone) {
// A simple module to replace `Backbone.sync` with *sessionStorage*-based
// persistence. Models are given GUIDS, and saved into a JSON object. Simple
// as that.

// Generate four random hex digits.
function S4() {
   return (((1+Math.random())*0x10000)|0).toString(16).substring(1);
};

// Generate a pseudo-GUID by concatenating random hexadecimal.
function guid() {
   return (S4()+S4()+"-"+S4()+"-"+S4()+"-"+S4()+"-"+S4()+S4()+S4());
};

function isObject(item) {
  return item === Object(item);
}

function contains(array, item) {
  var i = array.length;
  while (i--) if (array[i] === item) return true;
  return false;
}

function extend(obj, props) {
  for (var key in props) obj[key] = props[key]
  return obj;
}

function result(object, property) {
    if (object == null) return void 0;
    var value = object[property];
    return (typeof value === 'function') ? object[property]() : value;
}

// Our Store is represented by a single JS object in *localStorage* or *sessionStorage*. Create it
// with a meaningful name, like the name you'd give a table.
// window.Store is deprectated, use Backbone.WebStorage instead
Backbone.WebStorage = window.Store = function(name, storage) {
  if (!storage) {
    throw "Backbone.webStorage: Storage type not defined."
  }
  if( !this[storage] ) {
    throw "Backbone." + storage + ": Environment does not support " + storage + "."
  }
  this.name = name;
  this.storage = storage;
  this.serializer = {
    serialize: function(item) {
      return isObject(item) ? JSON.stringify(item) : item;
    },
    // fix for "illegal access" error on Android when JSON.parse is passed null
    deserialize: function (data) {
      return data && JSON.parse(data);
    }
  };
  var store = this[this.storage]().getItem(this.name);
  this.records = (store && store.split(",")) || [];
};

extend(Backbone.WebStorage.prototype, {

  // Save the current state of the **Store** to *localStorage* or *sessionStorage*.
  save: function() {
    this[this.storage]().setItem(this.name, this.records.join(","));
  },

  // Add a model, giving it a (hopefully)-unique GUID, if it doesn't already
  // have an id of it's own.
  create: function(model) {
    if (!model.id && model.id !== 0) {
      model.id = guid();
      model.set(model.idAttribute, model.id);
    }
    this[this.storage]().setItem(this._itemName(model.id), this.serializer.serialize(model));
    this.records.push(model.id.toString());
    this.save();
    return this.find(model);
  },

  // Update a model by replacing its copy in `this.data`.
  update: function(model) {
    this[this.storage]().setItem(this._itemName(model.id), this.serializer.serialize(model));
    var modelId = model.id.toString();
    if (!contains(this.records, modelId)) {
      this.records.push(modelId);
      this.save();
    }
    return this.find(model);
  },

  // Retrieve a model from `this.data` by id.
  find: function(model) {
    return this.serializer.deserialize(this[this.storage]().getItem(this._itemName(model.id)));
  },

  // Return the array of all models currently in storage.
  findAll: function() {
    var result = [];
    for (var i = 0, id, data; i < this.records.length; i++) {
      id = this.records[i];
      data = this.serializer.deserialize(this[this.storage]().getItem(this._itemName(id)));
      if (data != null) result.push(data);
    }
    return result;
  },

  // Delete a model from `this.data`, returning it.
  destroy: function(model) {
    this[this.storage]().removeItem(this._itemName(model.id));
    var modelId = model.id.toString();
    for (var i = 0, id; i < this.records.length; i++) {
      if (this.records[i] === modelId) {
        this.records.splice(i, 1);
      }
    }
    this.save();
    return model;
  },

  localStorage: function() {
    return localStorage;
  },

  sessionStorage: function() {
    return sessionStorage;
  },

  // Clear storage for specific collection.
  _clear: function() {
    var storage = this[this.storage](),
      itemRe = new RegExp("^" + this.name + "-");

    // Remove id-tracking item (e.g., "foo").
    storage.removeItem(this.name);

    // Match all data items (e.g., "foo-ID") and remove.
    for (var k in storage) {
      if (itemRe.test(k)) {
        storage.removeItem(k);
      }
    }

    this.records.length = 0;
  },

  // Size of storage.
  _storageSize: function() {
    return this[this.storage]().length;
  },

  _itemName: function(id) {
    return this.name+"-"+id;
  }

});

// localSync delegate to the model or collection's
// *localStorage* or *sessionStorage* property, which should be an instance of `Store`.
// window.Store.sync and Backbone.localSync is deprecated, use Backbone.WebStorage.sync instead
Backbone.WebStorage.sync = window.Store.sync = Backbone.localSync = function(method, model, options) {
  var store = result(model, 'localStorage') || result(model.collection, 'localStorage') || result(model, 'sessionStorage') || result(model.collection, 'sessionStorage');

  var resp, errorMessage;
  //If $ is having Deferred - use it.
  var syncDfd = Backbone.$ ?
    (Backbone.$.Deferred && Backbone.$.Deferred()) :
    (Backbone.Deferred && Backbone.Deferred());

  try {

    switch (method) {
      case "read":
        resp = model.id != undefined ? store.find(model) : store.findAll();
        break;
      case "create":
        resp = store.create(model);
        break;
      case "update":
        resp = store.update(model);
        break;
      case "delete":
        resp = store.destroy(model);
        break;
    }

  } catch(error) {
    if (error.code === 22 && store._storageSize() === 0)
      errorMessage = "Private browsing is unsupported";
    else
      errorMessage = error.message;
  }

  if (resp) {
    if (options && options.success) {
      if (Backbone.VERSION === "0.9.10") {
        options.success(model, resp, options);
      } else {
        options.success(resp);
      }
    }
    if (syncDfd) {
      syncDfd.resolve(resp);
    }

  } else {
    errorMessage = errorMessage ? errorMessage
                                : "Record Not Found";

    if (options && options.error)
      if (Backbone.VERSION === "0.9.10") {
        options.error(model, errorMessage, options);
      } else {
        options.error(errorMessage);
      }

    if (syncDfd)
      syncDfd.reject(errorMessage);
  }

  // add compatibility with $.ajax
  // always execute callback for success and error
  if (options && options.complete) options.complete(resp);

  return syncDfd && syncDfd.promise();
};

Backbone.ajaxSync = Backbone.sync;

Backbone.getSyncMethod = function(model, options) {
  var forceAjaxSync = options && options.ajaxSync;

  if(!forceAjaxSync && (result(model, 'localStorage') || result(model.collection, 'localStorage') || result(model, 'sessionStorage') || result(model.collection, 'sessionStorage'))) {
    return Backbone.localSync;
  }

  return Backbone.ajaxSync;
};

// Override 'Backbone.sync' to default to localSync,
// the original 'Backbone.sync' is still available in 'Backbone.ajaxSync'
Backbone.sync = function(method, model, options) {
  return Backbone.getSyncMethod(model, options).apply(this, [method, model, options]);
};

return Backbone.WebStorage;
}));
